### PR TITLE
MSC4108 support OIDC QR code login

### PR DESCRIPTION
**Pull request by t3chguy**

### CODE REVIEW

File: src/rendezvous/channels/MSC4108SecureChannel.ts
Comment (hughns) [inline src/rendezvous/channels/MSC4108SecureChannel.ts:214]: We should either remove this logging entirely, or only log the `type` field of the payload. Otherwise things like the secrets bundle get leaked.

```suggestion
        logger.debug(`=> {"type": ${JSON.stringify(payload.type)}, ...}`);
```

File: src/rendezvous/channels/MSC4108SecureChannel.ts
Comment (hughns) [inline src/rendezvous/channels/MSC4108SecureChannel.ts:234]: Same as before, we should either not log at all or only the `type`. Probably at debug log level too.

```suggestion
        logger.debug(`<= {"type": ${JSON.stringify(json.type)}, ...}`);
```

**Review by hughns — CHANGES_REQUESTED** [review]
Only minor changes suggested to prevent secrets leakage via logs.

**Review by hughns — APPROVED** [review]
The latest looks good to me (ignoring the build failure).

**Review by richvdh — COMMENTED** [review]
Sorry, I'm not going to be able to review this as it stands. Please break it up.

File: src/rendezvous/MSC4108SignInWithQR.ts
Comment (dbkr) [inline src/rendezvous/MSC4108SignInWithQR.ts:90]: This could do with some docs as a public export, as could the others above and in other files (the reason code exports are fairly self-evident).

File: src/rendezvous/transports/MSC4108RendezvousSession.ts
Comment (dbkr) [inline src/rendezvous/transports/MSC4108RendezvousSession.ts:184]: We should probably just disable this lint rule if we're going to ignore it anyway sometimes?

Comment (t3chguy) [reply src/rendezvous/transports/MSC4108RendezvousSession.ts:184]: We should probably specify `checkLoops: allExceptWhileTrue`

image

I will do that separately as it is set in https://github.com/matrix-org/eslint-plugin-matrix-org

Comment (t3chguy) [reply src/rendezvous/transports/MSC4108RendezvousSession.ts:184]: Looks like we can't use `allExceptWhileTrue` until we upgrade to eslint v9 which is blocked on some of the plugins we use becoming compatible

Comment (t3chguy) [reply src/rendezvous/transports/MSC4108RendezvousSession.ts:184]: https://github.com/matrix-org/eslint-plugin-matrix-org/issues/40

Comment (dbkr) [reply src/rendezvous/transports/MSC4108RendezvousSession.ts:184]: ah yep, I think florian was looking at upgrading this.

**Review by dbkr — APPROVED** [review]

**Comment by t3chguy** [comment]
> Sorry, I'm not going to be able to review this as it stands. Please break it up.

@richvdh you are only asked to review the crypto related bits to fulfil the CODEOWNERS requirement.

I have provided the links to the corresponding files for your convenience:

https://github.com/matrix-org/matrix-js-sdk/pull/4134/files#diff-1cdf58640fc7d7c465fab943126da6a0d5c9ab99301676278fdc362beea10c11
https://github.com/matrix-org/matrix-js-sdk/pull/4134/files#diff-55b0e7275b7510d86f5e201e7b377396032f2fe5f040c9681fb75d40213d2c01
https://github.com/matrix-org/matrix-js-sdk/pull/4134/files#diff-6eb6c51bc52963382d87f0e413ae91df69b9c8426285c367e27f0328db33360b
https://github.com/matrix-org/matrix-js-sdk/pull/4134/files#diff-ca8741e6064cda5d26243ec210815e74962e5617951c9ec99ef4877dad964b38
https://github.com/matrix-org/matrix-js-sdk/pull/4134/files#diff-890ab04f8e5e8b612d0069bf0c681ef35d1ef3810abba98205ee464f3c366577

PS this is already a lean version of the original PR, just with a plethora of tests and comments added which bloats it.

**Comment by richvdh** [comment]
Let's land #4227 and then see where we are

**Pull request merged**
